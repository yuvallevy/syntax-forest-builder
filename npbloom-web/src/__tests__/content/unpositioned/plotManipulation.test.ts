import { describe, expect, it } from 'vitest';
import {
  idMap, jsTreeMapRepr, NodeIndicatorInPlot, PlotCoordsOffset, set, StringSlice, TreeCoordsOffset,
  UnpositionedBranchingNode, UnpositionedNode, UnpositionedPlot, UnpositionedTerminalNode, UnpositionedTree
} from 'npbloom-core';

describe('plot manipulation', () => {
  const plot = new UnpositionedPlot(
    idMap({
      'cleo': new UnpositionedTree(
        'Cleo laughed.',
        idMap({
          's1': new UnpositionedBranchingNode('S', new TreeCoordsOffset(0, 5), set(['np1', 'vp1'])),
          'np1': new UnpositionedBranchingNode('NP', new TreeCoordsOffset(0, 0), set(['n1'])),
          'n1': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, 0), new StringSlice(0, 4)),
          'vp1': new UnpositionedTerminalNode('VP', new TreeCoordsOffset(0, 0), new StringSlice(5, 12)),
        }),
        new PlotCoordsOffset(0, 0),
      ),
      'alex': new UnpositionedTree(
        'Alex baked cookies.',
        idMap({
          's2': new UnpositionedBranchingNode('S', new TreeCoordsOffset(0, 5), set(['np2a', 'vp2'])),
          'np2a': new UnpositionedBranchingNode('NP', new TreeCoordsOffset(0, 0), set(['n2'])),
          'n2': new UnpositionedTerminalNode('N', new TreeCoordsOffset(0, 0), new StringSlice(0, 4)),
          'vp2': new UnpositionedBranchingNode('VP', new TreeCoordsOffset(0, 0), set(['v2', 'np2b'])),
          'v2': new UnpositionedTerminalNode('V', new TreeCoordsOffset(0, 0), new StringSlice(5, 10)),
          'np2b': new UnpositionedTerminalNode('NP', new TreeCoordsOffset(0, 0), new StringSlice(11, 18)),
        }),
        new PlotCoordsOffset(0, 0),
      ),
    }),
  );

  const changeOffset4PxUp = (node: UnpositionedNode): UnpositionedNode => {
    if (node instanceof UnpositionedTerminalNode) return new UnpositionedTerminalNode(
      node.label, new TreeCoordsOffset(node.offset.dTreeX, node.offset.dTreeY - 4), node.slice, node.triangle);
    if (node instanceof UnpositionedBranchingNode) return new UnpositionedBranchingNode(
      node.label, new TreeCoordsOffset(node.offset.dTreeX, node.offset.dTreeY - 4), node.children);
    throw Error("not part of test scope")
  };

  it('retrieves the tree ID and parent node ID of a single node', () => {
    expect(plot.getParentNodeIds(set([new NodeIndicatorInPlot('alex', 'vp2')])))
      .toStrictEqual(set([new NodeIndicatorInPlot('alex', 's2')]));
  });

  it('retrieves the tree ID and parent node ID of two sibling nodes', () => {
    expect(plot.getParentNodeIds(set([new NodeIndicatorInPlot('alex', 'np2a'), new NodeIndicatorInPlot('alex', 'vp2')])))
      .toStrictEqual(set([new NodeIndicatorInPlot('alex', 's2')]));
  });

  it('retrieve the tree ID and parent node IDs of two non-sibling nodes on the same tree', () => {
    expect(plot.getParentNodeIds(set([new NodeIndicatorInPlot('alex', 'np2a'), new NodeIndicatorInPlot('alex', 'n2')])))
      .toStrictEqual(set([new NodeIndicatorInPlot('alex', 's2'), new NodeIndicatorInPlot('alex', 'np2a')]));
  });

  it('retrieves no parent node IDs for a top-level node', () => {
    expect(plot.getParentNodeIds(set([new NodeIndicatorInPlot('alex', 's2')])))
      .toStrictEqual(set([]));
  });

  it('retrieves only one tree ID and parent node ID for two non-sibling nodes on the same tree, ' +
    'of which one is a top-level node', () => {
    expect(plot.getParentNodeIds(set([new NodeIndicatorInPlot('alex', 's2'), new NodeIndicatorInPlot('alex', 'n2')])))
      .toStrictEqual(set([new NodeIndicatorInPlot('alex', 'np2a')]));
  });

  it('transforms nodes across multiple trees on one plot', () => {
    expect(jsTreeMapRepr(plot.transformNodes(
      changeOffset4PxUp, set([new NodeIndicatorInPlot('cleo', 'np1'), new NodeIndicatorInPlot('alex', 's2')])
    ).trees)).toMatchSnapshot();
  });

  it('deletes nodes across multiple trees on one plot', () => {
    expect(jsTreeMapRepr(plot.deleteNodes(
      set([new NodeIndicatorInPlot('cleo', 's1'), new NodeIndicatorInPlot('cleo', 'np1'), new NodeIndicatorInPlot('alex', 'vp2')])
    ).trees)).toMatchSnapshot();
  });
});
