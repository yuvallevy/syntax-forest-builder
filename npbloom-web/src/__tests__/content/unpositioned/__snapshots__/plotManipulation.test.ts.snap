// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plot manipulation > deletes nodes across multiple trees on one plot 1`] = `
{
  "alex": {
    "nodes": {
      "n2": {
        "label": "N",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          0,
          4,
        ],
        "triangle": false,
      },
      "np2a": {
        "children": [
          "n2",
        ],
        "label": "NP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
      },
      "np2b": {
        "label": "NP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          11,
          18,
        ],
        "triangle": false,
      },
      "s2": {
        "children": [
          "np2a",
        ],
        "label": "S",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 5,
        },
      },
      "v2": {
        "label": "V",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          5,
          10,
        ],
        "triangle": false,
      },
    },
    "offset": {
      "dPlotX": 0,
      "dPlotY": 0,
    },
    "sentence": "Alex baked cookies.",
  },
  "cleo": {
    "nodes": {
      "n1": {
        "label": "N",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          0,
          4,
        ],
        "triangle": false,
      },
      "vp1": {
        "label": "VP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          5,
          12,
        ],
        "triangle": false,
      },
    },
    "offset": {
      "dPlotX": 0,
      "dPlotY": 0,
    },
    "sentence": "Cleo laughed.",
  },
}
`;

exports[`plot manipulation > transforms nodes across multiple trees on one plot 1`] = `
{
  "alex": {
    "nodes": {
      "n2": {
        "label": "N",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          0,
          4,
        ],
        "triangle": false,
      },
      "np2a": {
        "children": [
          "n2",
        ],
        "label": "NP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
      },
      "np2b": {
        "label": "NP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          11,
          18,
        ],
        "triangle": false,
      },
      "s2": {
        "children": [
          "np2a",
          "vp2",
        ],
        "label": "S",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 1,
        },
      },
      "v2": {
        "label": "V",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          5,
          10,
        ],
        "triangle": false,
      },
      "vp2": {
        "children": [
          "v2",
          "np2b",
        ],
        "label": "VP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
      },
    },
    "offset": {
      "dPlotX": 0,
      "dPlotY": 0,
    },
    "sentence": "Alex baked cookies.",
  },
  "cleo": {
    "nodes": {
      "n1": {
        "label": "N",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          0,
          4,
        ],
        "triangle": false,
      },
      "np1": {
        "children": [
          "n1",
        ],
        "label": "NP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": -4,
        },
      },
      "s1": {
        "children": [
          "np1",
          "vp1",
        ],
        "label": "S",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 5,
        },
      },
      "vp1": {
        "label": "VP",
        "offset": {
          "dTreeX": 0,
          "dTreeY": 0,
        },
        "slice": [
          5,
          12,
        ],
        "triangle": false,
      },
    },
    "offset": {
      "dPlotX": 0,
      "dPlotY": 0,
    },
    "sentence": "Cleo laughed.",
  },
}
`;
